doc/tutorial/security.dox

```diff
@@ -0,0 +1,247 @@
+ * @endlink which prevents this failure.
+ *
+ * @snippet tutorial.c Asymmetric key
+ * To use asymmetric key encryption keys, the publisher will need the
+ * public key (@ref _DPS_KeyCert.cert) of each recipient and each
+ * recipient will also need the private key (@ref
+ * _DPS_KeyCert.privateKey) and optional password (@ref
+ * _DPS_KeyCert.password).  The term recipient is used here instead of
+ * subscriber since a single recipient key may be used by multiple
+ * subscribers.
+ *
+ * For this example we are using only one recipient.  DPS supports
+ * Elliptic Curve Cryptography (ECC) certificates.  The certificate
+ * above is elided due to its size.
+ *
+ * @snippet tutorial.c Ephemeral asymmetric key handler
+ * The ephemeral key handler is used for both symmetric and asymmetric
+ * keys, so it must examine the incoming @c key type to determine what
+ * type of ephemeral key to return.  For @ref DPS_KEY_EC types, the
+ * ECC curve must also be examined.
+ *
+ * The ephemeral ECC key requested here is the ephemeral sender key.
+ * The actual key encryption key is derived from the sender and
+ * recipient ECC keys.
+ *
+ * In either case, the key handler creates a random key of the
+ * requested type (and curve) and returns it using DPS_SetKey().  If
+ * it cannot do that, it must return @ref DPS_ERR_MISSING.
+ *
+ * @section authenticating-the-message-sender Authenticating the message sender
+ *
+ * @snippet tutorial.c Creating a secure node with an authenticated sender
+ * Authenticating the sender requires that the @c keyId parameter
+ * be provided.  This is the key identifier of the public key (@ref
+ * _DPS_KeyCert.cert), private key (@ref _DPS_KeyCert.privateKey), and
+ * optional password (@ref _DPS_KeyCert.password) of the created node.
+ * This @c keyId value will be provided to the @ref DPS_KeyHandler
+ * when the node's own certificate is requested.
+ *
+ * @snippet tutorial.c Key handler
+ * The key handler implementation above supports all the mechanisms
+ * described so far.
+ *
+ * @snippet tutorial.c Ephemeral key handler
+ * The ephemeral key handler implementation above supports all the
+ * mechanisms described so far.
+ *
+ * @snippet tutorial.c Authenticating a publication
+ * @snippet tutorial.c Authenticating an acknowledgement
+ * In the publication or acknowledgement handler, the authenticated
+ * key identifier can be retrieved with
+ * DPS_PublicationGetSenderKeyId() or DPS_AckGetSenderKeyId().
+ *
+ * @section adding-access-control Adding access control
+ *
+ * Adding access control is a matter of combining the existing
+ * mechanisms to enforce any policies the application needs.  A simple
+ * example is shown below.
+ *
+ * @subsection defining-a-policy Defining a policy
+ *
+ * @snippet tutorial.c Defining a policy
+ * In the above we will allow @c alice to publish to topic @c a/b/c/d,
+ * @c bob to subscribe and acknowledge to topic @c a/b/c/d, and @c
+ * trudy to subscribe to topic @c a/b/c/d.
```

Comment:
it is probably because I am not actively working on DPS but what are topics?  when I see `a/b/c/d` this holds no meaning to me. also when the `BYTE_STR("alice")` when I looked up BYTE_STR the documentation stats it is a pre-shared key so the name in the example code is for illustration purposes right. This would actually be a generated public certificate key in a real program right?

does the ACL always have to end in a NULL entry?

Reply:
Yes, topic themselves are not covered in this work.  I assumed some background familiarity with pub/sub.  I'll mark this down for future work.

Reply:
BYTE_STR is just a macro used in the tutorial so I didn't have to type the string "alice", etc. in twice.  I used it anywhere in the API where it is expecting a buffer pointer and length.

ACLs are only in the tutorial code to show how you could tie the different APIs together to implement access control.

I'll see if I can make this a bit clearer.

Reply:
I'll put the introduction in now, I think a better long term choice is to fold the wiki (which has a good description of topics) into this documentation so it's all in one place and can be cross-referenced.